### ISSUE TITLE
add support `throw` keyword

### ISSUE THREAD

**Issue by sahandevs** [microsoft/TypeScript#40768]
## Search Terms

Error Type, throws, throw, throw keyword

## Explanation

add a `throw` keyword that when it reached in a type it reports a message/error.
it's useful when used with "Template literal types".

usage:
```ts
type RouteParams<T extends string> =
  T extends `/${infer Route}/:${infer Param}`
    ? { [Route]: { [Param]: string }}
    : throws `${T} doesn't have a "/:param" part`;

function addRoute<T extends string>(path: T, params: RouteParams<T>) ...

```

for now, we can only return types like this to inform the user why the input is incorrect:

```ts
type Error<T> = T & { error: true };

type RouteParams<T extends string> =
  T extends `/${infer Route}/:${infer Param}`
    ? { [Route]: { [Param]: string }}
    : Error<`${T} doesn't have a "/:param" part`>;

function addRoute<T extends string>(path: T, params: RouteParams<T>) ...

```

which doesn't show error lines and place of the error message is really indirect and the user may not found it/

## Checklist

My suggestion meets these guidelines:

* [x] This wouldn't be a breaking change in existing TypeScript/JavaScript code
* [x] This wouldn't change the runtime behavior of existing JavaScript code
* [x] This could be implemented without emitting different JS based on the types of the expressions
* [x] This isn't a runtime feature (e.g. library functionality, non-ECMAScript syntax with JavaScript output, etc.)
* [x] This feature would agree with the rest of [TypeScript's Design Goals](https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals).

**Comment by MartinJohns** [comment]
Duplicate of #23689.
